apps/nextjs/tests-e2e/tests/aila-chat/full-romans.test.ts

```diff
@@ -59,33 +66,109 @@ test(
 
     await test.step("Iterate through the fixtures", async () => {
       await page.waitForURL(/\/aila\/.+/);
-      await waitForGeneration(page, generationTimeout);
-      await expectSectionsComplete(page, 1);
-      await letUiSettle();
-
-      setFixture("roman-britain-2");
-      await continueChat(page);
-      await waitForGeneration(page, generationTimeout);
-      await expectSectionsComplete(page, 3);
-      await letUiSettle();
-
-      setFixture("roman-britain-3");
-      await continueChat(page);
-      await waitForGeneration(page, generationTimeout);
-      await expectSectionsComplete(page, 7);
-      await letUiSettle();
-
-      setFixture("roman-britain-4");
-      await continueChat(page);
-      await waitForGeneration(page, generationTimeout);
-      await expectSectionsComplete(page, 10);
-      await letUiSettle();
-
-      setFixture("roman-britain-5");
-      await continueChat(page);
-      await waitForGeneration(page, generationTimeout);
-      await expectSectionsComplete(page, 10);
 
+      let prevIteration = 0;
+      let prevSectionsComplete = 0;
+      const maxIterations = 20;
+
+      for (
+        let iterationCount = 1;
+        iterationCount <= maxIterations;
+        iterationCount++
+      ) {
+        setFixture(`roman-britain-${iterationCount}`);
+        await expectStreamingStatus(page, "RequestMade", { timeout: 5000 });
+
+        await waitForStreamingStatusChange(
+          page,
+          "RequestMade",
+          "Idle",
+          generationTimeout,
+        );
+
+        // Wait for the chat iteration to update
+        await page.waitForFunction(
+          ([prevIteration]) => {
+            const chatIteration = document.querySelector(
+              '[data-testid="chat-iteration"]',
+            );
+            return (
+              chatIteration &&
+              parseInt(chatIteration.textContent || "0", 10) >
+                (prevIteration ?? 0)
+            );
+          },
+          [prevIteration],
+          { timeout: generationTimeout },
+        );
+        // Update the previous iteration
+        const chatIteration = await page.textContent(
+          '[data-testid="chat-iteration"]',
+        );
+        prevIteration = parseInt(chatIteration || "0", 10);
+
+        // Get the current sections complete
+        const currSectionsComplete = await getSectionsComplete(page);
```

Comment:
This is more complex logic for checking the chat flow. Probably we move it to helpers?